addons/binding/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/internal/convention/homeassistant/HaID.java

```diff
@@ -64,62 +68,123 @@ public HaID(String mqttTopic) {
+    }
+
+    public HandlerConfiguration toHandlerConfiguration() {
+        String objectID = this.objectID;
+        if (StringUtils.isNotBlank(nodeID)) {
+            objectID = nodeID + "/" + objectID;
+        }
+
+        return new HandlerConfiguration(baseTopic, objectID);
+    }
+
+    public static HaID fromConfig(HandlerConfiguration config) {
+        String baseTopic = config.getBasetopic();
+        String objectID = config.getObjectid();
+        String nodeID = "";
+
+        if (StringUtils.contains(objectID, '/')) {
+            String[] parts = objectID.split("/");
+
+            if (parts.length != 2) {
+                throw new IllegalArgumentException(
+                        "Bad configuration. objectid must be <objectId> or <nodeId>/<objectId>!");
+            }
+            nodeID = parts[0];
+            objectID = parts[1];
         }
-        String[] groupParts = groupId.split("_");
-        if (groupParts.length != 2) {
-            throw new IllegalArgumentException("Channel needs a group ID with the pattern component_node!");
+        return new HaID(baseTopic, objectID, nodeID, "+");
+    }
+
+    public String getFallbackGroupId() {
+        StringBuilder str = new StringBuilder();
+
+        if (StringUtils.isNotBlank(nodeID)) {
+            str.append(nodeID).append('_');
         }
-        this.objectID = channel.getThingUID().getId();
-        this.nodeID = groupParts[1];
-        this.component = groupParts[0];
-        this.baseTopic = baseTopic;
+        str.append(objectID).append('_').append(component);
+        return str.toString();
     }
 
-    /**
-     * We map the HomeAssistant MQTT topic tree object to an ESH Thing.
-     */
-    public String getThingID() {
+    public String getComponent() {
+        return component;
+    }
+
+    public String getObjectID() {
         return objectID;
     }
 
-    /**
-     * The channel group type UID consists of all components of this object (object-id + node-id + component-id).
-     */
-    public String getChannelGroupTypeID() {
-        return objectID + "_" + component + nodeID;
+    public String getTopic(String suffix) {
+        StringBuilder str = new StringBuilder();
+
+        str.append(baseTopic).append('/').append(component).append('/');
+
+        if (StringUtils.isNotBlank(nodeID)) {
+            str.append(nodeID).append('/');
+        }
+        str.append(objectID);
+        if (StringUtils.isNotBlank(suffix)) {
+            str.append('/').append(suffix);
+        }
+
+        return str.toString();
     }
 
-    /**
-     * A channel type UID consists of all components of this object (object-id + node-id + component-id) and a
-     * channel-id on top.
-     */
-    public ChannelTypeUID getChannelTypeID(String channelID) {
-        return new ChannelTypeUID(MqttBindingConstants.BINDING_ID,
-                objectID + "_" + component + nodeID + "_" + channelID);
+    @Override
+    public int hashCode() {
```

Comment:
If you use basic types only like in this class, the java compiler will generate standard `hashCode` and `equals` methods.

Reply:
Am I missing something here?
`
public class Pojo {

    private final String name;

    public Pojo(String name) {
        this.name = name;
    }

    public static void main(String[] argv) {
        Pojo a = new Pojo("test");
        Pojo b = new Pojo("test");

        System.out.println("a.hashCode() = " + a.hashCode());
        System.out.println("b.hashCode() = " + b.hashCode());
        System.out.println("a.equals(b) =  " + a.equals(b));
    }
}
`
does not work.
And the default equals method inherited from Object is:
`
    public boolean equals(Object obj) {
        return (this == obj);
    }
`
which will only use the address of the object, not any members into account.

And I added the methods, because using `equals()` in the unit tests did not work.

Reply:
Hm. Alright. Different in C++.